jujugui/static/gui/src/app/state/state.js

```diff
@@ -0,0 +1,323 @@
+the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+if (typeof this.jujugui === 'undefined') {
+  this.jujugui = {};
+}
+
+const ROOT_RESERVED = [
+  'about', 'bigdata', 'docs', 'juju', 'login', 'logout', 'new', 'store'];
+const PROFILE_RESERVED = ['charms', 'issues', 'revenue', 'settings'];
+const PATH_DELIMETERS = new Map([['search', 'q'], ['user', 'u'], ['gui', 'i']]);
+const GUI_PATH_DELIMETERS = [
+  'account', 'applications', 'deploy', 'inspector', 'isv', 'machines'];
+
+/** Class representing the State of the Juju GUI */
+const State = class State {
+  /**
+    Create a new instance of the GUI state.
+    @param {Object} cfg - The configuration options for a new State instance.
+  */
+  constructor(cfg) {
+    /**
+      The baseURL to use when generating URLs. This value is required and
+      should be the full base path that the GUI is served from.
+      @property baseURL
+      @type {String}
+    */
+    if (!cfg.baseURL) {
+      throw 'baseURL must be provided.';
+    }
+    this.baseURL = cfg.baseURL;
+    /**
+      The list of dispatchers for the various components from the url.
+      @type {Object}
+    */
+    this.dispatchers = cfg.dispatchers;
+
+    if (cfg.seriesList && !Array.isArray(cfg.seriesList)) {
+      throw 'Series list must be an Array.';
+    }
+    /**
+      The list of possible distro 'series' for the store paths. ex)
+      trusty, precise, xenial. The GUI has a utils method `getSeriesList` that
+      is the central storage location for the series information. This Object
+      should be converted to an array and passed to this state when being
+      instantiated.
+      @type {Array}
+    */
+    this.seriesList = (function() {
+      if (!cfg.seriesList) {
+        console.log('No series list provided, using default list');
+        return ['bundle', 'precise', 'trusty', 'xenial'];
+      } else {
+        return cfg.seriesList.concat(['bundle']);
+      }
+    }());
+  }
+
+  /**
+    Takes a complete URL, strips off the supplied baseURL and extra slashes.
+    @param {String} url - The URL to sanitize.
+    @return {String} The sanitized path.
+  */
+  _getCleanPath(url) {
+    return url
+      // Strip the baseURL before parsing the sections.
+      .replace(this.baseURL, '')
+      // Strip the leading and trailing slashes off the url.
+      .replace(/^\/*/, '').replace(/\/*$/, '');
+  }
+
+  /**
+    Splits the URL up and generates a state object.
+    @param {String} url - The URL to turn into state.
+    @return {Object} The generated state object.
```

Comment:
Please update this to mention that the object includes the state and error fields, and what type each one is.